internal/content_coding_test.go

```diff
@@ -113,3 +118,112 @@ func TestStreamGzipDecode(t *testing.T) {
 
 	require.Equal(t, []byte("howdy"), b[:n])
 }
+
+func BenchmarkGzipEncodeDecode(b *testing.B) {
+	enc, err := NewGzipEncoder(-1)
+	require.NoError(b, err)
+	dec := NewGzipDecoder()
+
+	for n := 0; n < b.N; n++ {
+		payload, err := enc.Encode([]byte("howdy"))
+		require.NoError(b, err)
+
+		actual, err := dec.Decode(payload, maxDecompressionSize)
+		require.NoError(b, err)
+
+		require.Equal(b, "howdy", string(actual))
+	}
+}
+
+func BenchmarkGzipReuse(b *testing.B) {
+	enc, err := NewGzipEncoder(-1)
+	require.NoError(b, err)
+	dec := NewGzipDecoder()
+
+	payload, err := enc.Encode([]byte("howdy"))
+	require.NoError(b, err)
+
+	actual, err := dec.Decode(payload, maxDecompressionSize)
+	require.NoError(b, err)
+
+	require.Equal(b, "howdy", string(actual))
+
+	for n := 0; n < b.N; n++ {
+		payload, err = enc.Encode([]byte("doody"))
+		require.NoError(b, err)
+
+		actual, err = dec.Decode(payload, maxDecompressionSize)
+		require.NoError(b, err)
+
+		require.Equal(b, "doody", string(actual))
+	}
+}
```

Comment:
I think I caused some confusion here... Sorry for that. For each algorithm, I want to see a benchmark for the encoding/decoding cycle. As the `for` loop is what determines the measurement you should trim that one down as much as possible. Therefore, please remove the equality check.

Furthermore, we should **only** check the "reuse" case as we otherwise only include the buffer allocation time which does not make sense IMO. Anyway, your "non-reuse" test won't work as `b.N` is usually large and after `b.N > 0` you reuse the buffer anyway as you do not recreate the encoder/decoders.

It would maybe be good to increase the message size. Therefore my suggestion is to have the following for all algorithms (`gzip` and `identity` IIRC):
```suggestion
func BenchmarkGzipEncodeDecode(b *testing.B) {
	data := []byte(strings.Repeat("-howdy stranger-", 64))

	enc, err := NewGzipEncoder(-1)
	require.NoError(b, err)
	dec := NewGzipDecoder()

	payload, err := enc.Encode(data)
	require.NoError(b, err)
	actual, err := dec.Decode(payload, maxDecompressionSize)
	require.NoError(b, err)
	require.Equal(b, data, actual)

	for n := 0; n < b.N; n++ {
		p, err := enc.Encode(data)
		require.NoError(b, err)

		_, err = dec.Decode(p, maxDecompressionSize)
		require.NoError(b, err)
	}
}
```